wallet/src/account/output_cache/mod.rs

```diff
@@ -732,59 +745,143 @@ impl OutputCache {
+                            }
+                        }
+                    }
+                    WalletTx::Block(_) => {
+                        utils::debug_panic_or_log!("Cannot be block reward");
+                    }
+                };
+            }
+        }
+
+        // Remove all descendants of conflicting txs
+        let mut conflicting_tx_ids_with_descendants = vec![];
+        let mut conflicting_txs_with_descendants = vec![];
+
+        for conflicting_tx in conflicting_txs {
+            let tx_ids_to_rollback = self.remove_from_unconfirmed_descendants(conflicting_tx);
+
+            // Mark conflicting tx and its descendants as Conflicting and update OutputCache data accordingly
+            let mut tx_to_rollback_data = vec![];
+            for tx_id in tx_ids_to_rollback.iter().rev().copied() {
+                match self.txs.entry(tx_id.into()) {
+                    Entry::Occupied(mut entry) => match entry.get_mut() {
+                        WalletTx::Block(_) => {
+                            Err(WalletError::TransactionIdCannotMapToBlock(tx_id))
+                        }
+                        WalletTx::Tx(tx) => {
+                            conflicting_txs_with_descendants.push(WalletTx::Tx(tx.clone()));
+                            match tx.state() {
+                                TxState::Inactive(_) | TxState::InMempool(_) => {
+                                    tx.set_state(TxState::Conflicted(block_id));
+                                    tx_to_rollback_data.push(entry.get().clone());
+                                    Ok(())
+                                }
+                                TxState::Conflicted(..)
+                                | TxState::Abandoned
+                                | TxState::Confirmed(..) => {
+                                    Err(WalletError::CannotMarkTxAsConflictedIfInState(*tx.state()))
+                                }
+                            }
+                        }
+                    },
+                    Entry::Vacant(_) => {
+                        Err(WalletError::CannotFindDescendantTransactionWithId(tx_id))
+                    }
+                }?;
             }
+
+            for tx in tx_to_rollback_data {
+                self.rollback_tx_data(&tx)?;
+            }
+
+            conflicting_tx_ids_with_descendants.extend(tx_ids_to_rollback.into_iter());
         }
 
-        conflicting_txs
+        Ok(conflicting_tx_ids_with_descendants
             .into_iter()
-            .map(|tx_id| self.txs.get(tx_id).expect("must be present"))
-            .collect_vec()
+            .zip_eq(conflicting_txs_with_descendants)
+            .collect())
     }
 
-    fn uses_token(&self, unconfirmed_tx: &WalletTx, frozen_token_id: &TokenId) -> bool {
+    fn violates_frozen_token(&self, unconfirmed_tx: &WalletTx, frozen_token_id: &TokenId) -> bool {
         unconfirmed_tx.inputs().iter().any(|inp| match inp {
```

Comment:
[discussed] Mention in a comment that technically we should check the outputs too, but currently it's not possible to have a token in an output without having it in an input as well.